### PR TITLE
Add claimable balance support to simulation

### PR DESCRIPTION
**Pull request by sisuresh**
# Description

Add claimable balance support to our simulation mechanism.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v5.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/transactions/simulation/TxSimUtils.cpp [hunk 1]
```diff
@@ -109,6 +109,50 @@ generateScaledOfferID(OperationResult const& result, uint32_t partition)
...
+generateScaledClaimableBalanceID(Hash const& balanceID, uint32_t partition)
+{
+    if (partition == 0)
+    {
+        return balanceID;
+    }
+
+    auto scaledID = balanceID;
+
+    // XOR balanceID and partition
```
[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/transactions/simulation/TxSimUtils.cpp:144]: agree `SHA256(oldID | partition)` is the better way to do this

[the remark this answers is not in the document]
Comment (sisuresh) [reply src/transactions/simulation/TxSimUtils.cpp:144]: I didn't hash the partition because I didn't want to affect the performance of simulation. Is this not a concern? I would expect an additional SHA256 call for each scaled account/issuer on every op to be a non-negligible performance hit. 

The chances of having a collision with XOR in this case is extremely low (Jon's response to Marta's question . You might have to paste this link into slack first - https://stellarfoundation.slack.com/archives/C0367MGFL/p1595610545101900?thread_ts=1595465813.064400&cid=C0367MGFL).

If you think we should still go with the hash though, please let me know.

[the remark this answers is not in the document]
Comment (MonsieurNicolas) [reply src/transactions/simulation/TxSimUtils.cpp:144]: ah yes, this is fine. it is a hash already after all

File: src/transactions/simulation/TxSimUtils.cpp [hunk 2]
```diff
@@ -109,6 +109,50 @@ generateScaledOfferID(OperationResult const& result, uint32_t partition)
...
+
+    auto const& balanceID =
+        result.tr().createClaimableBalanceResult().balanceID().v0();
+    return generateScaledClaimableBalanceID(balanceID, partition);
+}
+
+Hash
+generateScaledClaimableBalanceID(Hash const& balanceID, uint32_t partition)
+{
+    if (partition == 0)
```
Comment (jonjove) [inline src/transactions/simulation/TxSimUtils.cpp:137]: Nitpick: I'm not sure how much this special case really matters. There is no problem with leaving it here, but it's probably marginal.

File: src/transactions/simulation/TxSimUtils.cpp [hunk 3]
```diff
@@ -176,6 +220,22 @@ generateScaledLiveEntries(std::vector<LedgerEntry>& entries,
             replaceIssuer(newEntry.data.offer().buying, partition);
             replaceIssuer(newEntry.data.offer().selling, partition);
             break;
+        case CLAIMABLE_BALANCE:
+        {
+            replaceIssuer(newEntry.data.claimableBalance().asset, partition);
+            mutateScaledAccountID(newEntry.data.claimableBalance().createdBy,
```
Comment (jonjove) [inline src/transactions/simulation/TxSimUtils.cpp:226]: This needs to be rebased, `createdBy` no longer exists.

Comment (sisuresh) [reply src/transactions/simulation/TxSimUtils.cpp]: Rebased and fixed

File: src/transactions/simulation/TxSimUtils.cpp (hunk 2 above)
Comment (sisuresh) [reply src/transactions/simulation/TxSimUtils.cpp]: Good point. Removed the case to make this simpler.

File: src/transactions/simulation/TxSimUtils.cpp [hunk 4]
```diff
@@ -177,6 +216,20 @@ generateScaledLiveEntries(std::vector<LedgerEntry>& entries,
...
+            replaceIssuer(newEntry.data.claimableBalance().asset, partition);
+            newEntry.data.claimableBalance().balanceID.v0() =
+                generateScaledClaimableBalanceID(
+                    newEntry.data.claimableBalance().balanceID.v0(), partition);
+
+            auto& claimants = newEntry.data.claimableBalance().claimants;
+            for (auto& claimant : claimants)
+            {
+                mutateScaledAccountID(claimant.v0().destination, partition);
+            }
```
Comment (jonjove) [inline src/transactions/simulation/TxSimUtils.cpp:230]: There is missing one other detail, which is more directly related to CAP-33 then to CAP-23. We aren't mutating the sponsorship information in the ledger entry. I really wanted to write this comment on line 186 but it won't let me.

While you are fixing this, can you also mutate the `signerSponsoringIDs` in account entry? Maybe make these two changes in a different commit.

Comment (sisuresh) [reply src/transactions/simulation/TxSimUtils.cpp:261]: Fixed

**Comment by jonjove** [comment]
r+ 00f8e3d
